tools/pubsys/src/repo.rs

```diff
@@ -0,0 +1,592 @@
+
+    info!(
+        "Using repo expiration policy from path: {}",
+        repo_args.repo_expiration_policy_path.display()
+    );
+    let expiration = RepoExpirationPolicy::from_path(&repo_args.repo_expiration_policy_path)
+        .context(error::Config)?;
+
+    let expiration_start_time = repo_args.release_start_time.unwrap_or(*DEFAULT_START_TIME);
+    let snapshot_expiration = expiration_start_time + expiration.snapshot_expiration;
+    let targets_expiration = expiration_start_time + expiration.targets_expiration;
+    let timestamp_expiration = expiration_start_time + expiration.timestamp_expiration;
+    info!(
+        "Repo expiration times:\n\tsnapshot:  {}\n\ttargets:   {}\n\ttimestamp: {}",
+        snapshot_expiration, targets_expiration, timestamp_expiration
+    );
+    editor
+        .snapshot_expires(snapshot_expiration)
+        .targets_expires(targets_expiration)
+        .timestamp_expires(timestamp_expiration);
+
+    // Add version   =^..^=   =^..^=   =^..^=   =^..^=
+
+    let seconds = Utc::now().timestamp();
+    let unsigned_seconds = seconds.try_into().expect("System clock before 1970??");
+    let version = NonZeroU64::new(unsigned_seconds).expect("System clock exactly 1970??");
+    debug!("Repo version: {}", version);
+    editor
+        .snapshot_version(version)
+        .targets_version(version)
+        .timestamp_version(version);
+
+    Ok(())
+}
+
+/// Builds an editor and manifest; will start from an existing repo if one is specified in the
+/// configuration
+fn build_editor_and_manifest(
+    repo_args: &RepoArgs,
+    infra_config: &InfraConfig,
+) -> Result<(RepositoryEditor, Manifest)> {
+    let repo_config = infra_config
+        .repo
+        .get(&repo_args.repo)
+        .context(error::MissingRepoConfig {
+            repo: &repo_args.repo,
+        })?;
+
+    // If repo section is defined, we have a matching repo name, and it has URLs defined, then try
+    // and load it, otherwise start a new repo
+    // (TODO: use Option.zip when stabilized)
+    let maybe_urls = repo_config.metadata_base_url.as_ref().and_then(|md_url| {
+        repo_config
+            .targets_url
+            .as_ref()
+            .map(|tgt_url| (md_url, tgt_url))
+    });
+    if let Some((metadata_url, targets_url)) = maybe_urls {
+        // Setup to load existing repo   =^..^=   =^..^=   =^..^=   =^..^=
+
+        info!("Found metadata and target URLs, loading existing repository");
+        let base_slash = if metadata_url.as_str().ends_with('/') {
+            ""
+        } else {
+            "/"
+        };
+        let metadata_url_str = format!(
+            "{}{}{}/{}",
+            metadata_url, base_slash, repo_args.variant, repo_args.arch
+        );
+        let metadata_url = Url::parse(&metadata_url_str).context(error::ParseUrl {
+            input: &metadata_url_str,
+        })?;
+        debug!("Using metadata url: {}", metadata_url);
```

Comment:
> It might improve readability to split out all the URL parsing into a helper, and only call this function with valid, formatted URLs.

We know the URLs coming from the config are valid by virtue of deserializing into `Url`.  The parsing here is effectively just checking that the variant and arch are valid as URL components, which is true unless the user set the wrong variables.

I'm not sure I understand what you're suggesting to split out.  We could have a function that takes the base URL and components and builds/parses a new one:  `fn metadata_url(metadata_base_url: Url, repo_args: RepoArgs) -> Result<Url>`  Calling that here would split out about 7 lines.  I'm not sure that it'd be worth the separation, but it's just stylistic and we could do that if you think it's clearer.  Or if I'm misinterpreting the request, please let me know.

> The "no URLs" case is pretty short and could go into run() or its own helper.

The goal of this function is abstracting where we get the editor and manifest - putting us in a consistent state.  I think moving the (large) conditional and the "new repo" case into the outer function makes things less consistent, and makes the outer function too large.

> I share the concern that the complexity of this function is pretty high. I understand what's going on, but it'd be nice to do something else with the nested editor_and_manifest_from_repo function. Maybe a macro?

Balancing this code was hard, and obviously not everyone agrees about the best way; there's just a bunch of straight-line actions we have to take.  :/

I think making the inner function a macro wouldn't help with complexity - it'd be basically the same number of lines, but add the complexity of macros, which are used sparingly in Rust.  Moving the inner function out to the top level would mean too many places to look for building an editor/manifest, and make you jump around the code even more when reading.  We tried to break things up into the main logical pieces, and keep each function consistent in achieving its single goal.  We thought hard about it and tried several alternatives and felt this was the best.  We'll continue to think, and welcome suggestions, and agree that this function is one of the more complex ones, but the relatively low-level logic of dealing with tough has to go somewhere, and this felt the best.

Reply:
We talked about this offline and got a better sense of the organization you were suggesting.  It worked out well and will be in the coming revision.